packages/fetchai/connections/http/connection.py

```diff
@@ -0,0 +1,384 @@
+                self.stopped = True
+                # self.thread.join()
+
+
+def HTTPHandlerFactory(channel: HTTPChannel):
+    class HTTPHandler(BaseHTTPRequestHandler):
+        def __init__(self, *args, **kwargs):
+            self._channel = channel
+            super(HTTPHandler, self).__init__(*args, **kwargs)
+
+        def do_HEAD(self):
+            self.send_response(200)
+            self.send_header("Content-type", "text/html")
+            self.end_headers()
+
+        def do_GET(self):
+            """Respond to a GET request."""
+            self.send_response(200)
+            self.send_header("Content-type", "text/html")
+            self.end_headers()
+
+            parsed_path = urlparse(self.path)
+            url = parsed_path.path
+            # url = parsed_path.geturl()
+            param = parse_qs(parsed_path.query)
+            param = json.dumps(param)
+
+            self._channel._process_request("GET", url, param)
+
+        def do_POST(self):
+            """Respond to a POST request."""
+            self.send_response(200)
+            self.send_header("Content-type", "text/html")
+            self.end_headers()
+
+            parsed_path = urlparse(self.path)
+            url = parsed_path.path
+            # url = parsed_path.geturl()
+            param = parse_qs(parsed_path.query)
+            param = json.dumps(param)
+
+            content_length = int(self.headers["Content-Length"])
+            body = self.rfile.read(content_length).decode()
+
+            self._channel._process_request("POST", url, param, body)
+
+    return HTTPHandler
+
+
+class HTTPConnection(Connection):
+    """Proxy to the functionality of the web RESTful API."""
+
+    def __init__(
+        self,
+        address: Address,
+        host: str = "",
+        port: int = 10000,
+        api_path: str = None,  # Directory path of the API YAML file.
+        *args,
+        **kwargs,
+    ):
+        """
+        Initialize a connection to an RESTful API.
+
+        :param api_spec_path: the directory path of the API specification YAML source file.
+        :param provider_addr: the provider IP address.
+        :param provider_port: the provider port.
+        :param connection_id: the identifier of the connection object.
+        :param restricted_to_protocols: the only supported protocols for this connection.
+        :param excluded_protocols: the excluded protocols for this connection.
+        """
+
+        # the following api_spec format checks will be in their own function check_api(api_spec)
+        if api_path is not None:
+            try:
+                with open(api_path, "r") as f:
+                    api_spec = yaml.safe_load(f)
+            except FileNotFoundError:
+                logger.error(
+                    "API specification YAML source file not found. Please double-check filename and path."
+                )
+                return
+        else:
+            api_spec = api_path
```

Comment:
All of this should be moved to the channel, just forward the `api_specification_path`

Reply:
I have moved the following to HTTPChannel.

Reply:
Update, it is now in _process_initial_api, but there is a check HTTPChannel that `api_spec_path` is not null.